### PR TITLE
Fix the dimensions mismatch bug for dataset with categorical columns

### LINKED ISSUE #270: ValueError: query data dimension must match training data dimension
while generating counterfactuals with 'genetic' method, the following error appears:

`ValueError: query data dimension must match training data dimension`

```
Traceback:
File "C:\Users\Anaconda3\envs\xyz\lib\site-packages\streamlit\script_runner.py", line 379, in _run_script
    exec(code, module.__dict__)
File "C:\Users\Downloads\myproject\counterfact_explain.py", line 102, in <module>
    desired_class=desired_var.item())
File "C:\Users\Anaconda3\envs\xyz\lib\site-packages\dice_ml\explainer_interfaces\explainer_base.py", line 102, in generate_counterfactuals
    **kwargs)
File "C:\Users\Anaconda3\envs\xyz\lib\site-packages\dice_ml\explainer_interfaces\dice_genetic.py", line 289, in _generate_counterfactuals
    sparsity_weight, diversity_weight, categorical_penalty, verbose)
File "C:\Users\Anaconda3\envs\xyz\lib\site-packages\dice_ml\explainer_interfaces\dice_genetic.py", line 208, in do_param_initializations
    query_instance, query_instance_df_dummies, verbose)
File "C:\Users\Anaconda3\envs\xyz\lib\site-packages\dice_ml\explainer_interfaces\dice_genetic.py", line 190, in do_cf_initializations
    indices = self.KD_tree.query(query_instance_df_dummies, num_queries)[1][0]
File "sklearn\neighbors\_binary_tree.pxi", line 1338, in sklearn.neighbors._kd_tree.BinaryTree.query
```

Dataset used : adult income dataset
Checked dimensions of both training data and query data, both are the same.
However, with 'random' method, it works perfectly.

Pls, help with this issue.

### PR DESCRIPTION
**Pull request by amit-sharma**
Fixes #270 
The error was because numerically encoded categorical variables were not considered categorical in the query_instance.

### CODE REVIEW

**Comment by amit-sharma** [comment]
Build 3.6 failed but all the pytest tests passed. the failure is in a subsequent github actions. Not sure why.

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 1]
```diff
@@ -82,7 +82,10 @@ def _generate_counterfactuals(self, query_instance, total_CFs, desired_range=Non
 
         # Prepares user defined query_instance for DiCE.
         query_instance_orig = query_instance.copy()
-        query_instance = self.data_interface.prepare_query_instance(query_instance=query_instance)
+        query_instance_orig = self.data_interface.prepare_query_instance(
```
Comment (gaugup) [inline dice_ml/explainer_interfaces/dice_KD.py:85]: Sorry, I don't understand the fix here? Could you elaborate?

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 2]
```diff
@@ -268,7 +268,10 @@ def _generate_counterfactuals(self, query_instance, total_CFs, initialization="k
 
         # Prepares user defined query_instance for DiCE.
         query_instance_orig = query_instance
-        query_instance = self.data_interface.prepare_query_instance(query_instance=query_instance)
+        query_instance_orig = self.data_interface.prepare_query_instance(
```
Comment (gaugup) [inline dice_ml/explainer_interfaces/dice_genetic.py:271]: Sorry, I don't understand the fix here? Could you elaborate?

**Comment by gaugup** [comment]
@amit-sharma could you add a unit test for this change as well as part of this PR?

File: dice_ml/explainer_interfaces/dice_KD.py (hunk 1 above)
Comment (amit-sharma) [reply dice_ml/explainer_interfaces/dice_KD.py:85]: The code was using two local variables, `query_instance` and `query_instance_orig`. The second variable was not passed through `prepare_query_instance`, hence the categorical features were not properly encoded.

**Comment by amit-sharma** [comment]
> @amit-sharma could you add a unit test for this change as well as part of this PR?

yes, added.

File: dice_ml/utils/helpers.py [hunk 3]
```diff
@@ -90,6 +95,11 @@ def load_custom_testing_dataset():
     return pd.DataFrame(data, columns=['Categorical', 'Numerical', 'Outcome'])
 
 
+def load_custom_vars_testing_dataset():
```
Comment (gaugup) [inline dice_ml/utils/helpers.py:98]: Shall we move this code to conftest.py? This shouldn't be a part of this package.

File: dice_ml/utils/helpers.py [hunk 4]
```diff
@@ -115,6 +125,26 @@ def load_custom_testing_dataset_regression():
     return pd.DataFrame(data, columns=['Categorical', 'Numerical', 'Outcome'])
 
 
+def save_custom_vars_dataset_model():
```
Comment (gaugup) [inline dice_ml/utils/helpers.py:128]: Same as above for this function.

File: dice_ml/explainer_interfaces/dice_KD.py (hunk 1 above)
Comment (gaugup) [reply dice_ml/explainer_interfaces/dice_KD.py:85]: Thanks for clarifying

File: dice_ml/utils/helpers.py (hunk 3 above)
Comment (amit-sharma) [reply dice_ml/utils/helpers.py:98]: Good point. I guess we can move some of the older functions also that are only used in test.

Comment (amit-sharma) [reply dice_ml/utils/helpers.py]: Done.

File: dice_ml/utils/helpers.py (hunk 4 above)
Comment (amit-sharma) [reply dice_ml/utils/helpers.py]: Done.

**Review by gaugup — APPROVED** [review]

**Pull request merged**